experiment/dash/dash.go

```diff
@@ -3,42 +3,134 @@ package dash
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
+	"time"
 
+	"github.com/dustin/go-humanize"
+	"github.com/montanaflynn/stats"
+	"github.com/neubot/dash/client"
+	neubotModel "github.com/neubot/dash/model"
 	"github.com/ooni/probe-engine/experiment"
 	"github.com/ooni/probe-engine/experiment/handler"
-	"github.com/ooni/probe-engine/experiment/mkevent"
-	"github.com/ooni/probe-engine/measurementkit"
 	"github.com/ooni/probe-engine/model"
 	"github.com/ooni/probe-engine/session"
 )
 
 const (
-	testName    = "dash"
-	testVersion = "0.7.0"
+	testName       = "dash"
+	testVersion    = "0.8.0"
+	defaultTimeout = 55 * time.Second
+	totalStep      = 15.0
 )
 
 // Config contains the experiment config.
-type Config struct {
-	// LogLevel is the MK log level. Empty implies "WARNING".
-	LogLevel string
+type Config struct{}
+
+// Simple contains the experiment total summary
+type Simple struct {
+	ConnectLatency  float64 `json:"Connect_latency"`
+	MedianBitrate   int64   `json:"Median_bitrate"`
+	MinPlayoutDelay float64 `json:"min_playout_delay"`
+}
+
+// TestKeys contains the test keys
+type TestKeys struct {
+	Simple `json:"simple"`
+
+	ClientResolver string `json:"client_resolver"`
+
+	// Failure is the failure string
+	Failure string `json:"failure"`
+
+	ReceiverData []neubotModel.ClientResults `json:"receiver_data"`
 }
 
 func measure(
 	ctx context.Context, sess *session.Session, measurement *model.Measurement,
 	callbacks handler.Callbacks, config Config,
 ) error {
-	settings := measurementkit.NewSettings(
-		"Dash", sess.SoftwareName, sess.SoftwareVersion,
-		sess.CABundlePath(), sess.ProbeASNString(), sess.ProbeCC(),
-		sess.ProbeIP(), sess.ProbeNetworkName(), config.LogLevel,
-	)
-	out, err := measurementkit.StartEx(settings, sess.Logger)
+	testkeys := &TestKeys{}
+	measurement.TestKeys = testkeys
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+	client := client.New(sess.SoftwareName, sess.SoftwareVersion)
+	// StartDownload starts the DASH download. It returns a channel where
+	// client measurements are posted, or an error.
+	ch, err := client.StartDownload(ctx)
 	if err != nil {
+		testkeys.Failure = err.Error()
 		return err
 	}
-	for ev := range out {
-		mkevent.Handle(sess, measurement, ev, callbacks)
+	callbacks.OnProgress(0, fmt.Sprintf("server: %s", client.FQDN))
+	var rates []float64
+	var frameReadyTime float64 = 0.0
+	var playTime float64 = 0.0
+	percentage := 0.0
+	step := 100 / (totalStep + 1) / 100
+	for results := range ch {
+		percentage += step
+		message := fmt.Sprintf(
+			"rate: %s/s speed: %s/s elapsed: %.2f s",
+			humanize.IBytes(uint64(results.Rate*1000/8)), // Rate is kbit in SI size
+			humanize.IBytes(uint64(float64(results.Received)/results.Elapsed)),
```

Comment:
I am always undecided whether to print `KiB/s` or `kbit/s`. Curious: why did you choose the former?

Reply:
Because every `2^10` is `1024`. And the standard says that `KiB` should be written for "every `1024 bytes` = `1 KB`". It seems to be no longer used everywhere.
(MacOS uses `10^2` and Windows uses `2^10`). But I don't feel good about using `10^2` instead of `2^10`.
https://en.wikipedia.org/wiki/Binary_prefix#Deviation_between_powers_of_1024_and_powers_of_1000

Reply:
Here's my rule of thumb: either kbit/s (base 10) or KiB/s (base 2). The former is the unit in which they sell you the connection. The latter is the unit in which you download/upload stuff that matters. There is clearly a tension. I've seen tools showing the speed in kbit/s and tools showing it in KiB/s. My personal preference, when writing speed tests, has always been using kbit/s (or Mbit/s, or whatever) under the argument that other speed tests do the same.

Reply:
Okay, so let's consider the `Conventional current`. :P